added Transactions creating and handling functionality

igorya7v opened the pull request.

Commit:
added Transactions creating and handling functionality

compscidr commented on src/main/java/SuperPeer.java:
```diff
@@ -21,35 +26,44 @@ public static void main(String[] args) {
+        
     }
-    
-    public void handleTransactionPacket(MeshManager.RightMeshEvent e) {
```
This shouldn't have been taken out - this is how all of the transaction requests should be handled...this function should parse the request types and execute them in the ethereum network.

igorya7v replied:
I moved it to the TransactionManager

compscidr replied:
Yep just saw that - was about to update comment ...haha

compscidr commented on rm-ethereum.conf:
```diff
@@ -0,0 +1,14 @@
+"maxgasLimit":"1000000000",
+"debugInfo":"false",
+"rpcAddress":"http://localhost:8545",
+"channelAddr":"0x0f1736FA3eB54BD8C234725BD658201DA0B4190D",
```
Just wondering why the channel address is specified here? Aren't we creating the channels and getting new ones? Or is this the channel contract address?

igorya7v replied:
Yep, it's the Contract address.
I will update it's name, to channelContractAddress.
I also want to all Settings to the library as they are the same on both sides.
We will do it after the merge with David changes.

comey replied:
I my MR, these information has been included in a JSON file included in our library.

compscidr commented on rm-ethereum.conf:
```diff
@@ -0,0 +1,14 @@
+"debugInfo":"false",
+"rpcAddress":"http://localhost:8545",
+"channelAddr":"0x0f1736FA3eB54BD8C234725BD658201DA0B4190D",
+"tokenAddr":"0x0fC373426c87F555715E6fE673B07Fe9E7f0E6e7",
```
Is this our deployed TGE contract on the Kovan network? Or one of David's test contracts?

comey replied:
I think this is just my test contracts

igorya7v replied:
This is the David's test contracts yet.

fractalic replied:
for consistency, if the above is changed to `channelContractAddress`, I think it makes sense to change this to `tokenContractAddress`

compscidr replied:
The tokenContract and channelContract are two different things.

compscidr replied:
Nevermind I get what you mean lol. I haven't had enough coffee yet.

compscidr commented on src/main/java/SuperPeer.java:
```diff
@@ -21,35 +26,44 @@ public static void main(String[] args) {
-            msg = c.readLine();
+            //msg = c.readLine();
+            try {
+                msg = br.readLine();
```
Here we can probably add something that lets the person running the superpeer close a channel (good for testing at this stage if we can't have the phones issue close channels yet) (don't need to do in this MR, but suggestion for later)

FrazerSeymour commented on src/main/java/SuperPeer.java:
```diff
@@ -1,18 +1,23 @@
  
 public class SuperPeer {
 
+    //TODO: Add logger instead of system.out
```
Would `MeshUtility.Log()` do the trick? It _also_ just uses System.out for now but we can possibly configure the backend there in the future.

FrazerSeymour replied:
Thinking on this - probably not. I don't think that function was intended to be a logging solution for programs _using_ the library, just the library itself.

igorya7v replied:
Yep, we will add some logger later.

FrazerSeymour commented on src/main/java/SuperPeer.java:
```diff
@@ -21,35 +26,44 @@ public static void main(String[] args) {
+
         //test send on the TRANSACTION PORT
-        mm.sendDataReliable(mm.getUuid(), MeshUtility.TRANSACTION_PORT, "test".getBytes());
+        //mm.sendDataReliable(mm.getUuid(), MeshUtility.TRANSACTION_PORT, "test".getBytes());
```
Commented out code here and below should be deleted.

FrazerSeymour commented on src/main/java/ether/Http.java:
```diff
@@ -0,0 +1,80 @@
+            response.close();
+            httpClient.close();
+
+            temp=new BasicResponseHandler().handleResponse(response);
```
Our Checkstyle plugin would probably get mad at a bunch of the formatting in here. @compscidr - are we enforcing those on this project? Should we?

compscidr replied:
Yes, I'd prefer if we ran everything through checkstyle. This means we need to add it to our githooks and also figure out a way to add to intellij or whatever IDE ppl are using for the non-android devving.

FrazerSeymour commented on src/main/java/ether/Http.java:
```diff
@@ -0,0 +1,80 @@
+                    executionResult=jobj.get(key);
+                }
+            }
+        }catch (UnsupportedEncodingException e) {
```
You could probably collapse this whole thing into
```
catch (UEE | CPE | IOE | ...  e) {
    System.out.println(e.class.getStringOrWhatever() + ": " + e.messageOrWhatever);
}
```

FrazerSeymour commented on src/main/java/ether/JSON.java:
```diff
@@ -0,0 +1,147 @@
+/**
+ * The JSON wrapper
+ */
+public final class JSON {
```
Is this  JSON wrapper? It appears to be sending messages on the mesh. A more descriptive name might help.

FrazerSeymour replied:
Wrapping the creation of JSON objects into a class might not be a bad idea, but that logic shouldn't have a "sending over the network" side effect.

FrazerSeymour commented on src/main/java/ether/TransactionsManager.java:
```diff
@@ -0,0 +1,663 @@
+
+        try {
+            executor.shutdown();
+            executor.awaitTermination(5, TimeUnit.SECONDS);
```
Should this be configured somewhere? Or at least moved to a constant/documented why 5 seconds?

FrazerSeymour commented on src/main/java/ether/TransactionsManager.java:
```diff
@@ -0,0 +1,663 @@
+     * Processes the Get All request from Client-Remote Peer
+     * @param sourceId
+     */
+    private void processGetAllRequest(MeshID sourceId) {
```
This is a _really_ long function - can any chunks of it be moved into their own functions?

FrazerSeymour commented on src/main/java/ether/EtherClient.java:
```diff
@@ -0,0 +1,719 @@
+                "\"params\":[" +
+                "{" +
+                "\"to\":\""+Settings.TOKEN_ADDRESS+"\"," +
+                "\"data\":\""+"0x" + new String(org.apache.commons.codec.binary.Hex.encodeHex(functionBytes))+"\"" +
```
Importing that function might make this slightly more readable.

fractalic reviewed (changes requested):
Formatting quibble: some lines are very long and should probably be wrapped.

FrazerSeymour commented on src/main/java/ether/JSON.java:
```diff
@@ -0,0 +1,147 @@
+/**
+ * The JSON wrapper
+ */
+public final class JSON {
```
One other thing - we've been using Protobuf for the data payloads we send around RightMesh as they're small and nice to work with. It would be worth considering replacing JSON with Protobuf in the future.

igorya7v replied:
This is the plan))

compscidr replied:
Yes - we discussed this last week - since the microraiden-java project used json, it is easier if we continue to use it....like igor mentioned later on we will switch it to protobuf.

fractalic commented on src/main/java/ether/TransactionsManager.java:
```diff
@@ -0,0 +1,663 @@
+    private static final int NUM_OF_THREADS = 100;
+
+    private MeshManager meshManager;
+    private LinkedBlockingQueue<MeshManager.MeshTransactionEvent> transactionsQueue = new LinkedBlockingQueue<>();
```
consider using abstract interface BlockingQueue as the type unless there is an explicit dependency on functions not defined in the interface. Same for next line.

fractalic commented:
In the TransactionsManager, where is the Settings object defined? I see references to Settings.STUFF, but I can't figure out where this object is defined, and how it gets included in TransactionsManager.

FrazerSeymour commented on src/main/java/ether/TransactionsManager.java:
```diff
@@ -0,0 +1,663 @@
+
+    private static volatile TransactionsManager instance = null;
+
+    public static TransactionsManager getInstance(MeshManager mm) {
```
Why is this a singleton? It appears to be only created in the one place in the codebase.

Singletons are tricky to unit test and we're definitely going to want them for this codebase  in the future.

igorya7v replied:
It's a good question. I just wanted to ensure that there will be only one instance of this Object, no matter how and where we are going to use it. As, this Object manages the Transactions queue and it's thread, and register itself as a Transaction handler to the MeshManager. Therefore, if somehow multiple instances will be created it will be a mess. If we can ensure that only one instance will be created during it's use, than it's OK to change it to be not a Singleton.

FrazerSeymour replied:
Makes sense. I suppose we can leave it for now and change it if/when we right tests for this codebase.

Are there other places in the code where you can envision us needing another instance of this class?

igorya7v replied:
Not right now, lets reconsider how we would like to use it later. We could switch easily to the none singleton version, if we would like to.

fractalic reviewed (commented):
superPeerToClientChannel and clientToSuperPeerChannel:

names like `outgoingChannel` and `incomingChannel` or even just `outgoing` or `incoming` would be shorter and easier to read.

FrazerSeymour commented on src/main/java/ether/TransactionsManager.java:
```diff
@@ -0,0 +1,663 @@
+
+
+        //Calculate Client-->SuperPeer hash
+        byte[] clientToSuperPeerHash = EtherUtility.getChannelHash(sourceId.toString(), ownMeshId.toString());
```
Is this duplicate code to line 351? If so these should probably be functions.

FrazerSeymour replied:
Looks like a few other of these checks are duplicated in a few places as well (e.g. fetching `clientToSuperPeerChannel` from the blockchain) - moving the duplicate checks to their own functions will make this a lot more readable.

FrazerSeymour commented on src/main/java/ether/EtherClient.java:
```diff
@@ -0,0 +1,719 @@
+
+        CallTransaction.Function channels = Settings.CHANNEL_CONTRACT.getByName("channels");
+        byte [] functionBytes=channels.encode(keyInBytes);
+        String requestString = "{\"method\":\"eth_call\"," +
```
If you use this request string in multiple places, it might help to make one functino with the string in it and pass in the addresses and data as parameters.

sriks1 commented on src/main/java/ether/EtherClient.java:
```diff
@@ -0,0 +1,719 @@
+        byte[] keyInBytes= EtherUtility.getChannelHash(senderAddress, receiverAddress);
+
+        CallTransaction.Function channels = Settings.CHANNEL_CONTRACT.getByName("channels");
+        byte [] functionBytes=channels.encode(keyInBytes);
```
this line might throw an exception if there is no channel set between the send and receive address. should we check first if the channel exist?

sriks1 commented on src/main/java/ether/EtherClient.java:
```diff
@@ -0,0 +1,719 @@
+            System.out.println("Waiting for network to mine transactions ... ");
+        }
+
+        waitingForTransaction(transactionId, httpAgent);
```
are we running the whole transaction stuff in a separate thread - just in case if it takes too much time to mine?

igorya7v replied:
You are right Sri, each transaction is executed on a different Thread.

//execute on a separate Thread
executor.execute(() -> processTransaction(event));

Commit:
Changes to address the MR comments

FrazerSeymour reviewed (approved):
Looks good. :+1:

Commit:
- Transactions Hashmap was removed.
- Transactions processing changed from simultaneous to synchronous.
- Added general http request construction method.
- Added Close channels from command line functionality.
- Integration with David's TransactionManager and Transactins HashMaps.

comey commented on rm-ethereum.conf:
```diff
@@ -0,0 +1,14 @@
+{
```
Can super access these values in JSON file with the library? Because some of these values in the config file are the same with JSON files.

igorya7v replied:
Yes, we will do it after the demo.

comey commented on rm-ethereum.conf:
```diff
@@ -9,6 +9,6 @@
 "channelABI":"[{\"constant\":true,\"inputs\":[],\"name\":\"challenge_period\",\"outputs\":[{\"name\":\"\",\"type\":\"uint32\"}],\"payable\":false,\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"version\",\"outputs\":[{\"name\":\"\",\"type\":\"string\"}],\"payable\":false,\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"name\":\"_receiver_address\",\"type\":\"address\"} ...
 "appendingZerosForETH":"1000000000000000000",
 "appendingZerosForTKN":"1000000000000000000",
-"maxDepositBits":"192",
```
The maxDepositBits is the number of bits when we convert maximum deposit into binary. If you change the value from 192 to 10, it means a person can put 1024 Wei of token at maximum. So, we do not need to change this value, since 2 to the power 192 is defined in payment channel contract.

igorya7v replied:
Got it.
Do we want to add our Maximum deposit limit.

comey commented on src/main/java/ether/EtherClient.java:
```diff
@@ -30,238 +29,280 @@ private EtherClient() {
+            return null;
+        }
+
+        String depositStr = response.substring(0, response.length() / 2);
```
What is the format of the response? Could you please add a comment for the reason of dividing by 2?

igorya7v replied:
The response contains the Channel struct, struct is returned as an array of bytes
Channel struct contains 2 members - initial deposit and the open block number.
I will add a comment in the code.

comey commented on src/main/java/ether/TransactionsManager.java:
```diff
@@ -437,308 +487,105 @@ private void processGetAllRequest(MeshID sourceId) {
-        //Get the Client-->SuperPeer channel object
-        EtherUtility.PaymentChannel inChannel = getChannelFromMap(sourceId.toString(), ownMeshId.toString());
+            //validate the balance in both pairs (balanceProofSig and closingSig)
+            if(balanceProofSig.right.equals(closingSig.right)) {
```
For an in-channel, if the balance values are not the same between balanceProof and closingHash, the super peer can recreate a new closingHash signature with the balance in balanceProof, then the channel can be closed.

comey replied:
We cannot guarantee the balances are the same, since DATA and DATA_ACK may be in the transmission path.

comey replied:
Similar to the in-channel case, when the super peer wants close an out-channel, if the two balances are not the same, the super peer need to recreate a balanceProof signature to match the balance in closingHash. Then, the channel can also be closed. You can keep a log in the output but the channel needs to be closed.

igorya7v replied:
done

FrazerSeymour commented on src/main/java/SuperPeer.java:
```diff
@@ -1,55 +1,98 @@
-            mm.stop();
-            return;
+            Thread.sleep(200);
+        } catch (Exception ex) {
```
Catching Exception is a bad habit. Checkstyle might yell at you for it too.

igorya7v replied:
done

FrazerSeymour commented on src/main/java/SuperPeer.java:
```diff
@@ -1,55 +1,98 @@
+    }
+
+    private void processInput(String msg) {
+        if(msg.equals("exit")) {
```
Would it make sense to make this a constant?

igorya7v replied:
done

Commit:
- If needed, added ClosingSig and BalanceProof regeneration when closing channels.

Commit:
- added commands constants

Commit:
- Disabled gas estimation for Open Channel transaction.
  Currently, we don't know why it doesn't work.
  Will be fixed later.
- Added some prints for the demo.

Commit:
cleanup

FrazerSeymour commented:
Is the commented code coming back? Or should it be totally removed?

igorya7v commented:
It's just a temporally commented out, we need to fix it.

Commit:
added call to System.exit to terminate on exit command

igorya7v committed:
Merge branch 'master' into RM-303

compscidr reviewed (commented):
Going to push ahead with merging - we should probably do a more details review after the demo
